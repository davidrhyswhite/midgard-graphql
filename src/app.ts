import { Config } from 'apollo-server-core';
import { ApolloServer as Server } from 'apollo-server';
import { ApolloServer as Lambda } from 'apollo-server-lambda';
import { loadConfigSync } from 'graphql-config';
import schema from './schema';
import MidgardAPI from './datasources/midgard';
import { DataSources } from './types/context';

const config = loadConfigSync({ rootDir: './' });

interface ApolloConstructable<T> {
  new (args: Config): T;
}
type ServerType = Server | Lambda;

export function createServer(App: ApolloConstructable<ServerType>): ServerType {
  return new App({
    schema,
    playground: {
      config,
    },
    introspection: true,
    dataSources: (): DataSources => ({
      midgardAPI: new MidgardAPI(),
    }),
  });
}
