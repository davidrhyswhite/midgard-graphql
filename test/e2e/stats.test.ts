import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import nock from 'nock';
import app from '../../src/app';

const GET_STATS = gql`
  query getStats {
    stats {
      dailyActiveUsers
      monthlyActiveUsers
      totalUsers
      dailyTx
      monthlyTx
      totalTx
      totalVolume24hr
      totalVolume
      totalStaked
      totalDepth
      totalEarned
      poolCount
      totalAssetBuys
      totalAssetSells
      totalStakeTx
      totalWithdrawTx
    }
  }
`;

test('returns a stats type from application', async () => {
  const { query } = createTestClient(app);

  nock('http://midgard.hostname.local').get('/stats/path').reply(200, {
    dailyActiveUsers: 'dailyActiveUsers',
    dailyTx: 'dailyTx',
    monthlyActiveUsers: 'monthlyActiveUsers',
    monthlyTx: 'monthlyTx',
    poolCount: 'poolCount',
    totalAssetBuys: 'totalAssetBuys',
    totalAssetSells: 'totalAssetSells',
    totalDepth: 'totalDepth',
    totalEarned: 'totalEarned',
    totalStakeTx: 'totalStakeTx',
    totalStaked: 'totalStaked',
    totalTx: 'totalTx',
    totalUsers: 'totalUsers',
    totalVolume: 'totalVolume',
    totalVolume24hr: 'totalVolume24hr',
    totalWithdrawTx: 'totalWithdrawTx',
  });

  // run query against the server and snapshot the output
  const res = await query({ query: GET_STATS });
  expect(res).toMatchSnapshot();
});
