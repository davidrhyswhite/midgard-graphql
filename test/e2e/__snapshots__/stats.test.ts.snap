// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns a stats type from application 1`] = `
Object {
  "data": Object {
    "stats": Object {
      "dailyActiveUsers": "dailyActiveUsers",
      "dailyTx": "dailyTx",
      "monthlyActiveUsers": "monthlyActiveUsers",
      "monthlyTx": "monthlyTx",
      "poolCount": "poolCount",
      "totalAssetBuys": "totalAssetBuys",
      "totalAssetSells": "totalAssetSells",
      "totalDepth": "totalDepth",
      "totalEarned": "totalEarned",
      "totalStakeTx": "totalStakeTx",
      "totalStaked": "totalStaked",
      "totalTx": "totalTx",
      "totalUsers": "totalUsers",
      "totalVolume": "totalVolume",
      "totalVolume24hr": "totalVolume24hr",
      "totalWithdrawTx": "totalWithdrawTx",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
